In [1]:
import pandas as pd
import numpy as np

In [2]:
import matplotlib.pyplot as plt
import seaborn as sns

In [3]:
from pathlib import Path
import warnings
warnings.filterwarnings("ignore")

In [4]:
# adjusting display settings customize
pd.set_option("display.max_columns",None)
pd.set_option("display.float_format", "{:,.2f}".format)

In [5]:
data_dir = Path("E:/DATA_ANALYSIS_PROJECTS/Amazon_Sales/dataset")

In [6]:
files = list(data_dir.glob("*.csv"))
[file.name for file in files]

['Amazon Sale Report.csv',
 'Cloud Warehouse Compersion Chart.csv',
 'Expense IIGF.csv',
 'International sale Report.csv',
 'May-2022.csv',
 'P  L March 2021.csv',
 'Sale Report.csv']

In [7]:
file_name = "Amazon Sale Report.csv"
df_amz= pd.read_csv(data_dir/file_name)

print("File Shape :: ", df_amz.shape)
df_amz.head()

File Shape ::  (128975, 24)


,index,Order ID,Date,Status,Fulfilment,Sales Channel,ship-service-level,Style,SKU,Category,Size,ASIN,Courier Status,Qty,currency,Amount,ship-city,ship-state,ship-postal-code,ship-country,promotion-ids,B2B,fulfilled-by,Unnamed: 22
0,0,405-8078784-5731545,04-30-22,Cancelled,Merchant,Amazon.in,Standard,SET389,SET389-KR-NP-S,Set,S,B09KXVBD7Z,NaN,0,INR,647.62,MUMBAI,MAHARASHTRA,"400,081.00",IN,NaN,False,Easy Ship,NaN
1,1,171-9198151-1101146,04-30-22,Shipped - Delivered to Buyer,Merchant,Amazon.in,Standard,JNE3781,JNE3781-KR-XXXL,kurta,3XL,B09K3WFS32,Shipped,1,INR,406.00,BENGALURU,KARNATAKA,"560,085.00",IN,Amazon PLCC Free-Financing Universal Merchant ...,False,Easy Ship,NaN
2,2,404-0687676-7273146,04-30-22,Shipped,Amazon,Amazon.in,Expedited,JNE3371,JNE3371-KR-XL,kurta,XL,B07WV4JV4D,Shipped,1,INR,329.00,NAVI MUMBAI,MAHARASHTRA,"410,210.00",IN,IN Core Free Shipping 2015/04/08 23-48-5-108,True,NaN,NaN
3,3,403-9615377-8133951,04-30-22,Cancelled,Merchant,Amazon.in,Standard,J0341,J0341-DR-L,Western Dress,L,B099NRCT7B,NaN,0,INR,753.33,PUDUCHERRY,PUDUCHERRY,"605,008.00",IN,NaN,False,Easy Ship,NaN
4,4,407-1069790-7240320,04-30-22,Shipped,Amazon,Amazon.in,Expedited,JNE3671,JNE3671-TU-XXXL,Top,3XL,B098714BZP,Shipped,1,INR,574.00,CHENNAI,TAMIL NADU,"600,073.00",IN,NaN,False,NaN,NaN


In [8]:
df_amz.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 128975 entries, 0 to 128974
Data columns (total 24 columns):
 #   Column              Non-Null Count   Dtype  
---  ------              --------------   -----  
 0   index               128975 non-null  int64  
 1   Order ID            128975 non-null  object 
 2   Date                128975 non-null  object 
 3   Status              128975 non-null  object 
 4   Fulfilment          128975 non-null  object 
 5   Sales Channel       128975 non-null  object 
 6   ship-service-level  128975 non-null  object 
 7   Style               128975 non-null  object 
 8   SKU                 128975 non-null  object 
 9   Category            128975 non-null  object 
 10  Size                128975 non-null  object 
 11  ASIN                128975 non-null  object 
 12  Courier Status      122103 non-null  object 
 13  Qty                 128975 non-null  int64  
 14  currency            121180 non-null  object 
 15  Amount              121180 non-nul

In [9]:
audit = pd.DataFrame({
    "column": df_amz.columns,
    "dtype": df_amz.dtypes,
    "null_%": df_amz.isna().mean()*100,
    "unique_value": df_amz.nunique()
}).sort_values("null_%", ascending=False)

audit

,column,dtype,null_%,unique_value
fulfilled-by,fulfilled-by,object,69.55,1
promotion-ids,promotion-ids,object,38.11,5787
Unnamed: 22,Unnamed: 22,object,38.03,1
currency,currency,object,6.04,1
Amount,Amount,float64,6.04,1410
Courier Status,Courier Status,object,5.33,3
ship-country,ship-country,object,0.03,1
ship-city,ship-city,object,0.03,8955
ship-state,ship-state,object,0.03,69
ship-postal-code,ship-postal-code,float64,0.03,9459


# Column Audit & Cleaning Decisions

-- This section documents column-level decisions made after profiling the dataset.
The goal is to ensure business correctness, analytical usefulness, and downstream compatibility with SQL, Power BI, and Machine Learning models.

## Columns to Drop (No Business Value)

| Column Name   | Decision | Reason                                                               |
| ------------- | -------- | -------------------------------------------------------------------- |
| `Unnamed: 22` | Drop     | System-generated CSV artifact with no analytical or business meaning |
| `index`       | Drop     | Technical row index, not part of business grain or identifiers       |


Rationale:
Removing non-business columns improves model clarity and avoids confusion in BI and SQL layers.

# Columns with High Missing Values (Retained with Transformation)
| Column Name     | Null % | Decision           | Reason                                                                |
| --------------- | ------ | ------------------ | --------------------------------------------------------------------- |
| `fulfilled-by`  | ~69.5% | Keep → derive flag | Null implies non-Amazon fulfilment; presence implies Amazon-fulfilled |
| `promotion-ids` | ~38.1% | Keep → derive flag | Null indicates no promotion; useful for discount impact analysis      |

## Derived Features Planned:

is_amazon_fulfilled (Boolean)

has_promotion (Boolean)

# Financial Columns (Preserved for Integrity)
| Column Name | Null % | Decision | Reason                                            |
| ----------- | ------ | -------- | ------------------------------------------------- |
| `Amount`    | ~6.0%  | Keep     | Missing values reflect cancelled or unpaid orders |
| `currency`  | ~6.0%  | Keep     | Required for financial consistency and validation |


    Rows with missing revenue will not be dropped to preserve the complete order lifecycle.

# Logistics & Fulfilment Columns
| Column Name      | Null % | Decision | Reason                                                    |
| ---------------- | ------ | -------- | --------------------------------------------------------- |
| `Courier Status` | ~5.3%  | Keep     | Missing values indicate unassigned or cancelled shipments |


# Geographic Columns
| Column Name        | Decision          | Reason                                             |
| ------------------ | ----------------- | -------------------------------------------------- |
| `ship-postal-code` | Convert to string | Postal codes are identifiers, not numerical values |
| `ship-city`        | Keep              | Used for city-level analysis                       |
| `ship-state`       | Keep              | Used for regional aggregation                      |
| `ship-country`     | Keep              | Required for domestic vs international analysis    |


# Core Business Columns (Retained As-Is)

These columns are complete, consistent, and central to sales analytics:

Order ID

Date (to be converted to datetime)

Status

Sales Channel

Fulfilment

ship-service-level

SKU

ASIN

Category

Style

Size

Qty

B2B

Grain Definition:
One row represents one product line item within an order.

# 🎯 Summary

Columns are dropped only when they add no business value

Missing values are interpreted using business context, not blindly removed

Decisions are aligned with:

SQL star schema design

Power BI performance

Machine Learning feature engineering

In [10]:
df= df_amz.copy()

In [11]:
drop_cols = ["Unnamed: 22" , "index"]
df.drop(columns=drop_cols, inplace=True)

In [12]:
df['Date'] = pd.to_datetime(df["Date"], errors="coerce")

In [13]:
df["order_year"] = df["Date"].dt.year
df["order_month"] = df["Date"].dt.month
df["order_day"] = df["Date"].dt.day

In [14]:
df["has_promotion"] = df["promotion-ids"].notna()

df["is_amazon_fulfilled"] = df["fulfilled-by"].notna()

df["Courier Status"] = df["Courier Status"].fillna("Not Assigned")

df["ship-postal-code"] = df["ship-postal-code"].astype("Int64").astype(str)

df["is_revenue_recorded"] = df["Amount"].notna()


In [15]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 128975 entries, 0 to 128974
Data columns (total 28 columns):
 #   Column               Non-Null Count   Dtype         
---  ------               --------------   -----         
 0   Order ID             128975 non-null  object        
 1   Date                 128975 non-null  datetime64[ns]
 2   Status               128975 non-null  object        
 3   Fulfilment           128975 non-null  object        
 4   Sales Channel        128975 non-null  object        
 5   ship-service-level   128975 non-null  object        
 6   Style                128975 non-null  object        
 7   SKU                  128975 non-null  object        
 8   Category             128975 non-null  object        
 9   Size                 128975 non-null  object        
 10  ASIN                 128975 non-null  object        
 11  Courier Status       128975 non-null  object        
 12  Qty                  128975 non-null  int64         
 13  currency      

In [16]:
df.isna().mean().sort_values(ascending=False).head(10)

fulfilled-by     0.70
promotion-ids    0.38
currency         0.06
Amount           0.06
ship-state       0.00
ship-city        0.00
ship-country     0.00
Sales Channel    0.00
Order ID         0.00
Date             0.00
dtype: float64

# DATA MODELING (FACT–DIMENSION DESIGN)

In [17]:
dim_product = (
    df[["SKU","ASIN","Category","Style","Size"]]
    .drop_duplicates()
    .reset_index(drop=True)
)

dim_product.head()

,SKU,ASIN,Category,Style,Size
0,SET389-KR-NP-S,B09KXVBD7Z,Set,SET389,S
1,JNE3781-KR-XXXL,B09K3WFS32,kurta,JNE3781,3XL
2,JNE3371-KR-XL,B07WV4JV4D,kurta,JNE3371,XL
3,J0341-DR-L,B099NRCT7B,Western Dress,J0341,L
4,JNE3671-TU-XXXL,B098714BZP,Top,JNE3671,3XL


In [18]:
dim_geography =(
    df[["ship-city", "ship-state", "ship-country", "ship-postal-code"]]
    .drop_duplicates()
    .reset_index(drop=True)
)

dim_geography.head()

,ship-city,ship-state,ship-country,ship-postal-code
0,MUMBAI,MAHARASHTRA,IN,400081
1,BENGALURU,KARNATAKA,IN,560085
2,NAVI MUMBAI,MAHARASHTRA,IN,410210
3,PUDUCHERRY,PUDUCHERRY,IN,605008
4,CHENNAI,TAMIL NADU,IN,600073


In [19]:
df.columns.tolist()

for col in df.columns:
    print(repr(col))


'Order ID'
'Date'
'Status'
'Fulfilment'
'Sales Channel '
'ship-service-level'
'Style'
'SKU'
'Category'
'Size'
'ASIN'
'Courier Status'
'Qty'
'currency'
'Amount'
'ship-city'
'ship-state'
'ship-postal-code'
'ship-country'
'promotion-ids'
'B2B'
'fulfilled-by'
'order_year'
'order_month'
'order_day'
'has_promotion'
'is_amazon_fulfilled'
'is_revenue_recorded'


In [20]:
dim_channel = (
    df[["Sales Channel ", "Fulfilment", "ship-service-level"]]
    .drop_duplicates()
    .reset_index(drop=True)
)

dim_channel.head()


,Sales Channel,Fulfilment,ship-service-level
0,Amazon.in,Merchant,Standard
1,Amazon.in,Amazon,Expedited
2,Amazon.in,Amazon,Standard
3,Non-Amazon,Amazon,Standard


In [21]:
dim_date = (
    df[["Date", "order_year", "order_month", "order_day"]]
    .drop_duplicates()
    .assign(
        month_name=lambda x: x["Date"].dt.month_name(),
        quarter=lambda x: x["Date"].dt.to_period("Q").astype(str)
    )
    .reset_index(drop=True)
)

dim_date.head()


,Date,order_year,order_month,order_day,month_name,quarter
0,2022-04-30,2022,4,30,April,2022Q2
1,2022-04-29,2022,4,29,April,2022Q2
2,2022-04-28,2022,4,28,April,2022Q2
3,2022-04-27,2022,4,27,April,2022Q2
4,2022-04-26,2022,4,26,April,2022Q2


In [22]:
fact_sales = df[
    [
        "Order ID",
        "Date",
        "SKU",
        "ASIN",
        "Qty",
        "Amount",
        "currency",
        "Status",
        "is_revenue_recorded",
        "has_promotion",
        "is_amazon_fulfilled",
        "B2B"
    ]
].copy()

fact_sales.head()


,Order ID,Date,SKU,ASIN,Qty,Amount,currency,Status,is_revenue_recorded,has_promotion,is_amazon_fulfilled,B2B
0,405-8078784-5731545,2022-04-30,SET389-KR-NP-S,B09KXVBD7Z,0,647.62,INR,Cancelled,True,False,True,False
1,171-9198151-1101146,2022-04-30,JNE3781-KR-XXXL,B09K3WFS32,1,406.00,INR,Shipped - Delivered to Buyer,True,True,True,False
2,404-0687676-7273146,2022-04-30,JNE3371-KR-XL,B07WV4JV4D,1,329.00,INR,Shipped,True,True,False,True
3,403-9615377-8133951,2022-04-30,J0341-DR-L,B099NRCT7B,0,753.33,INR,Cancelled,True,False,True,False
4,407-1069790-7240320,2022-04-30,JNE3671-TU-XXXL,B098714BZP,1,574.00,INR,Shipped,True,False,False,False


In [23]:
print("Fact rows:", fact_sales.shape)
print("Product dim:", dim_product.shape)
print("Geo dim:", dim_geography.shape)
print("Channel dim:", dim_channel.shape)
print("Date dim:", dim_date.shape)


Fact rows: (128975, 12)
Product dim: (7200, 5)
Geo dim: (18230, 4)
Channel dim: (4, 3)
Date dim: (91, 6)


## “I modeled the dataset into a star schema with a sales fact table and normalized dimensions to support SQL analytics, Power BI dashboards, and ML pipelines.”

In [27]:
# Python to DB connect

from sqlalchemy import create_engine

engine = create_engine(
    "postgresql+psycopg://postgres:Pritam_1612@localhost:5432/amazon_dw"
)

In [28]:
with engine.connect() as conn:
    print("PostgreSQL connected successfully")



PostgreSQL connected successfully


In [31]:
print(dim_product.columns.tolist())


['SKU', 'ASIN', 'Category', 'Style', 'Size']


In [32]:
dim_product.columns = dim_product.columns.str.lower()


In [33]:
print(dim_product.columns.tolist())


['sku', 'asin', 'category', 'style', 'size']


In [34]:
dim_product.to_sql(
    "dim_product",
    engine,
    schema="analytics",
    if_exists="append",
    index=False
)


-1

In [35]:
for df_dim in [dim_geography, dim_channel, dim_date]:
    df_dim.columns = df_dim.columns.str.lower()


In [ ]:
dim_geography.to_sql(
    "dim_geography",
    engine,
    schema="analytics",
    if_exists="append",
    index=False
)
